topo/node/drivenets/drivenets.go

```diff
@@ -0,0 +1,303 @@
+			Image:          config.Image,
+			Command:        config.Command[0],
+			Args:           config.Args,
+			Env:            node.ToEnvVar(config.Env),
+			ConfigPath:     config.ConfigPath,
+			ConfigFile:     config.ConfigFile,
+			InitImage:      config.InitImage,
+			Ports:          ports,
+			InterfaceCount: len(nodeSpec.Interfaces) + 1,
+			InitSleep:      int(config.Sleep),
+			Resources:      node.ToResourceRequirements(nodeSpec.Constraints),
+			Labels: nodeSpec.Labels,
+		},
+	}
+	if config.Cert != nil {
+		switch tls := config.Cert.Config.(type) {
+		case *tpb.CertificateCfg_SelfSigned:
+			dut.Spec.TLS = &cdnosv1.TLSSpec{
+				SelfSigned: &cdnosv1.SelfSignedSpec{
+					CommonName: tls.SelfSigned.CommonName,
+					KeySize:    int(tls.SelfSigned.KeySize),
+				},
+			}
+		}
+	}
+
+	cs, err := clientFn(n.RestConfig)
+	if err != nil {
+		return fmt.Errorf("failed to get kubernetes client: %v", err)
+	}
+	if _, err := cs.CdnosV1alpha1().Cdnoss(n.Namespace).Create(ctx, dut, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("failed to create cdnos: %v", err)
+	}
+	return nil
+}
+
+func (n *Node) Status(ctx context.Context) (node.Status, error) {
+		if n.Impl.Proto.Model != modelCdnos {
+		return node.StatusUnknown, fmt.Errorf("invalid model specified")
+	}
+	return n.cdnosStatus(ctx)
+}
+
+func (n *Node) cdnosStatus(ctx context.Context) (node.Status, error) {
+	cs, err := clientFn(n.RestConfig)
+	if err != nil {
+		return node.StatusUnknown, err
+	}
+	got, err := cs.CdnosV1alpha1().Cdnoss(n.Namespace).Get(ctx, n.Name(), metav1.GetOptions{})
+	if err != nil {
+		return node.StatusUnknown, err
+	}
+	switch got.Status.Phase {
+	case cdnosv1.Running:
+		return node.StatusRunning, nil
+	case cdnosv1.Failed:
+		return node.StatusFailed, nil
+	case cdnosv1.Pending:
+		return node.StatusPending, nil
+	default:
+		return node.StatusUnknown, nil
+	}
+}
+
+func (n *Node) Delete(ctx context.Context) error {
+	if n.Impl.Proto.Model != modelCdnos {
+		return fmt.Errorf("unknown model")
+	}
+	return n.cdnosDelete(ctx)
+}
+
+func (n *Node) cdnosDelete(ctx context.Context) error {
+	cs, err := clientFn(n.RestConfig)
+	if err != nil {
+		return err
+	}
+	return cs.CdnosV1alpha1().Cdnoss(n.Namespace).Delete(ctx, n.Name(), metav1.DeleteOptions{})
+}
+
+func (n *Node) ResetCfg(ctx context.Context) error {
+	log.Info("ResetCfg is a noop.")
+	return nil
+}
+
+func (n *Node) ConfigPush(context.Context, io.Reader) error {
+	return status.Errorf(codes.Unimplemented, "config push is not implemented using gNMI to configure device")
+}
+
+func (n *Node) GenerateSelfSigned(context.Context) error {
+	return status.Errorf(codes.Unimplemented, "certificate generation is not supported")
+}
+
+func cdnosDefaults(pb *tpb.Node) *tpb.Node {
+	if pb.Config == nil {
+		pb.Config = &tpb.Config{}
+	}
+	if pb.Config.Image == "" {
+		pb.Config.Image = "pr-registry.dev.drivenets.net/cdnos_pr_61485:19.1.0.1_priv.61485.1ca7037408b5f7a975949e0912d743498b1f49b5"
```

Comment:
this should probably point to a moving image tag, is this open source btw?

Reply:
what do you mean by "a moving immage tag" ?
no the image is not open source, need to contact Drivenets

Reply:
The default image tag should look like it points to a production release, I imagine something like `whatever.com/cdnos:latest`, even if that is not publicly accessible -- the one shown looks like it's an internal development build. For that dev work, you can always override the image in the config section of the node in the topology with the `image` statement.

IMO, the reason for wanting a more "final" looking image tag is that a user is more likely to have such an image tag loaded into the container runtime than the dev build one, and `:latest` is the convention for finding the most recent one.

I, as the eventual customer of this image, could similarly use that `image` statement to point KNE at a copy of a specific version of the cDNOS image in our internal registry.

Reply:
ok, understood, i setup a more official path